Fix #847

Pr0methean opened the pull request.

Pr0methean committed:
Fix #847

Boats are now consumed when used. Regression test added. (Also adds a
commented-out attempt to test this for projectile items; I'll finish
that part in a separate PR if I can do so without help.)

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into bug847

Pr0methean committed:
Remove an unnecessary unchecked cast

Pr0methean committed:
Inline

Pr0methean committed:
Fix projectile tests

Pr0methean committed:
Unused imports

Pr0methean committed:
Make "projectile" a field

In case subclasses need it.

aramperes commented on src/test/java/net/glowstone/block/itemtype/ItemBoatTest.java:
```diff
@@ -0,0 +1,72 @@
+        verify(world, times(1)).spawn(any(Location.class), eq(Boat.class));
+        assertEmpty(inventory.getItemInMainHand());
+    }
+}
```
Add newline at end of file

aramperes commented on src/main/java/net/glowstone/block/itemtype/ItemBoat.java:
```diff
@@ -47,6 +48,7 @@ private void placeBoat(GlowPlayer player) {
             location.setYaw(player.getLocation().getYaw());
             Boat boat = targetBlock.getWorld().spawn(location, Boat.class);
             boat.setWoodType(woodType);
+            InventoryUtil.consumeHeldItem(player, holding);
```
Don't you have to forcefully set the slot after changing the amount? IIRC it doesn't update the inventory until you use setSlot().

Pr0methean replied:
Added a test case in `ItemProjectileTest` to confirm that it's written back to the GlowPlayerInventory even in the non-zero-remaining case. As for updating the client, that's handled the next time the player is pulsed, by `GlowPlayer` lines 863-5.

Pr0methean committed:
Add tests for stack of two

Confirms that consumeHeldItem isn't operating on a detached copy when it
decrements the count.

Pr0methean committed:
Add testRightClickBlock

Pr0methean committed:
Refactor

Pr0methean committed:
Style fix

mastercoms reviewed (changes requested):
I tested it in-game, and the boat is not consumed.

Pr0methean committed:
Merge branch 'dev' into bug847

Pr0methean commented:
I'll resume work on this once #829 is merged, since I'll likely need to extend `GlowEntityTest` from that PR.

Pr0methean committed:
Newline at end of file

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into bug847

# Conflicts:
#	src/test/java/net/glowstone/block/itemtype/ItemProjectileTest.java
#	src/test/java/net/glowstone/block/itemtype/ItemTypeTest.java

Pr0methean committed:
Merge remote-tracking branch 'origin/bug847' into bug847

# Conflicts:
#	src/test/java/net/glowstone/block/itemtype/ItemProjectileTest.java
#	src/test/java/net/glowstone/block/itemtype/ItemTypeTest.java

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into bug847

Pr0methean committed:
Empty commit to force merge of #829

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into bug847

mastercoms commented:
Closing due to https://github.com/GlowstoneMC/Glowstone/issues/847#issuecomment-370089545.
